pkg/controller/node/nodecontroller.go

```diff
@@ -232,14 +239,47 @@ func NewNodeController(
 		largeClusterThreshold:       largeClusterThreshold,
 		unhealthyZoneThreshold:      unhealthyZoneThreshold,
 		zoneStates:                  make(map[string]zoneState),
+		runTaintManager:             runTaintManager,
 	}
 	nc.enterPartialDisruptionFunc = nc.ReducedQPSFunc
 	nc.enterFullDisruptionFunc = nc.HealthyQPSFunc
 	nc.computeZoneStateFunc = nc.ComputeZoneState
 
 	podInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    nc.maybeDeleteTerminatingPod,
-		UpdateFunc: func(_, obj interface{}) { nc.maybeDeleteTerminatingPod(obj) },
+		AddFunc: func(obj interface{}) {
+			nc.maybeDeleteTerminatingPod(obj)
+			pod := obj.(*v1.Pod)
+			if nc.runTaintManager {
+				nc.taintManager.PodUpdated(nil, pod)
```

Comment:
The PodUpdated function is non trivial and potential blocking (it's putting an element to channel).
As a result, this cannot be called from handler function, as it can potentially block all controllers which is super bad.

You should use the following pattern:
AddFunc: "enqueue the pod that should be processed"
UpdateFunc: "enqueue the pod that should be processed"
DeleteFunc: "enqueue the pod that should be processed"

And then have a worker thread (1 or more) that will be processing those enqueued pods.

This is pretty well documented here:
https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md

[To be honest, this should already be the case previously].

Thank you @deads2k for pointing me to this doc few days ago.